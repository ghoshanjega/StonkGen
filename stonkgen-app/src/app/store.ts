import { configureStore, ThunkAction, Action,getDefaultMiddleware } from '@reduxjs/toolkit';
import authenticationReducer from '../features/authentication/AuthenticationSlice';
import stocksListReducer from '../features/stocksList/StocksListSlice';
import cartReducer from '../features/cart/CartSlice';

export const store = configureStore({
  reducer: {
    authentication : authenticationReducer,
    stocksList: stocksListReducer,
    cart: cartReducer
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware({
    serializableCheck: false
}),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
