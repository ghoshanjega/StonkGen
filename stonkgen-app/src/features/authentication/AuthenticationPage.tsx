import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { Field, Form, Formik } from "formik";
import { login, signup, selectUser, selectDisplayLogin, toggleDisplayLogin, selectError } from "./AuthenticationSlice";
import { useHistory } from "react-router-dom";

export const AuthenticationPage = () => {
  const dispatch = useDispatch();
  const user = useSelector(selectUser);
  const displayLogin = useSelector(selectDisplayLogin);
  const error = useSelector(selectError);

  const history = useHistory();

  if (user) {
    history.push("/");
  }

  const wrapInLoginToggle = (text: string, disabled: boolean = false) => {
    return (
      <button className={`btn btn-lg btn-link`} data-test={text} disabled={disabled} onClick={() => dispatch(toggleDisplayLogin())}>
        {text}
      </button>
    );
  };

  return (
    <div className="container">
      <div className="row">
        <div className="col-md-5 mx-auto p-5">
          <div className="d-flex justify-content-between mb-3">
            <div>{displayLogin ? wrapInLoginToggle("Login", true) : wrapInLoginToggle("Login")}</div>
            <div>{!displayLogin ? wrapInLoginToggle("Sign Up", true) : wrapInLoginToggle("Sign Up")}</div>
          </div>
          <Formik
            initialValues={{ userName: "", password: "", fullName: "" }}
            onSubmit={(values, actions) => {
              actions.setSubmitting(false);
              displayLogin ? dispatch(login(values)) : dispatch(signup(values));
            }}
          >
            {({ isSubmitting }) => (
              <Form>
                {!displayLogin && (
                  <div className="form-group">
                    <label htmlFor="fullName">Full Name</label>
                    <Field type="text" name="fullName" className="form-control" />
                  </div>
                )}
                <div className="form-group">
                  <label htmlFor="userName">User Name</label>
                  <Field type="text" name="userName" className="form-control" />
                </div>
                <div className="form-group mb-3">
                  <label htmlFor="password">Password</label>
                  <Field type="text" name="password" className="form-control" />
                </div>

                <div className="col-md-12 text-center ">
                  <button type="submit" className=" btn btn-block mybtn btn-primary tx-tfm" disabled={isSubmitting}>
                    Login
                  </button>
                </div>
                {error && <p className="text-danger my-3 text-center">{error.message}</p>}
              </Form>
            )}
          </Formik>
        </div>
      </div>
    </div>
  );
};
