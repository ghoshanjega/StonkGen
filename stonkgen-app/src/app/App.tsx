import React, { useContext } from "react";
import { useSelector } from "react-redux";
import { BrowserRouter, Route, Switch } from "react-router-dom";
import { Header } from "../components/Header";

import { AuthenticationPage } from "../features/authentication/AuthenticationPage";
import { selectUser } from "../features/authentication/AuthenticationSlice";
import { CartPage } from "../features/cart/CartPage";
import { StocksListPage } from "../features/stocksList/StocksListPage";

import "./App.css";

function App() {
  const user = useSelector(selectUser);
  if (!user) {
    return (
      <div>
        <AuthenticationPage />
      </div>
    );
  }
  return (
    <div className="wrapper">
      <BrowserRouter>
        <Header />
        <Switch>
          <Route path="/orders">
            <CartPage />
          </Route>
          <Route path="/login">
            <AuthenticationPage />
          </Route>
          <Route path="/">
            <StocksListPage />
          </Route>
        </Switch>
      </BrowserRouter>
    </div>
  );
}

export default App;
