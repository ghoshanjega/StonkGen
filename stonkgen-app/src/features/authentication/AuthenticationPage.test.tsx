import React from "react";
import { AuthenticationPage } from "./AuthenticationPage";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import { store } from "../../app/store";


describe("Authentication Component", () => {
  let component: renderer.ReactTestRenderer;
  beforeEach(() => {
    store.dispatch = jest.fn();

    component = renderer.create(
      <Provider store={store}>
        <AuthenticationPage />
      </Provider>
    );
  });
  it("should render the login snapshot", () => {
    expect(component).toMatchSnapshot();
  });

  it("renders username field", () => {
    expect(component.root.findByProps({ htmlFor: "userName" }).children).toEqual(["User Name"]);
  });

  it("renders login button", () => {
    expect(component.root.findByProps({ type: "submit" }).children).toEqual(["Login"]);
  });

  it("should dispatch an action on login button click", async () => {
    await renderer.act(async () => {
      component.root.findByType("form").props.onSubmit();
    });
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(expect.any(Function));
  });

  it("should dispatch an action on signup button click", async () => {
    renderer.act( () => {
      component.root.findByProps({"data-test":"Sign Up"}).props.onClick();
    })
    expect(store.dispatch).toHaveBeenCalledTimes(1);
  });
});
