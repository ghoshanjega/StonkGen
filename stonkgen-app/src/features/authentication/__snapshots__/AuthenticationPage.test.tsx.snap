// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authentication Component should render the login snapshot 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-md-5 mx-auto p-5"
    >
      <div
        className="d-flex justify-content-between mb-3"
      >
        <div>
          <button
            className="btn btn-lg btn-link"
            data-test="Login"
            disabled={true}
            onClick={[Function]}
          >
            Login
          </button>
        </div>
        <div>
          <button
            className="btn btn-lg btn-link"
            data-test="Sign Up"
            disabled={false}
            onClick={[Function]}
          >
            Sign Up
          </button>
        </div>
      </div>
      <form
        action="#"
        onReset={[Function]}
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            htmlFor="userName"
          >
            User Name
          </label>
          <input
            className="form-control"
            name="userName"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group mb-3"
        >
          <label
            htmlFor="password"
          >
            Password
          </label>
          <input
            className="form-control"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="col-md-12 text-center "
        >
          <button
            className=" btn btn-block mybtn btn-primary tx-tfm"
            disabled={false}
            type="submit"
          >
            Login
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
